src/fsharp/PrettyNaming.fs

```diff
@@ -588,20 +588,20 @@ module public FSharp.Compiler.PrettyNaming
     let IllegalCharactersInTypeAndNamespaceNames = [| '.'; '+'; '$'; '&'; '['; ']'; '/'; '\\'; '*'; '\"'; '`'  |]
 
     /// Determines if the specified name is a valid name for an active pattern.
-    let IsActivePatternName (nm: string) =
-        let nameLen = nm.Length
-        // The name must start and end with '|'
-        (nm.IndexOf '|' = 0) &&
-        (nm.LastIndexOf '|' = nameLen - 1) &&
+    let IsActivePatternName (name: string) =
         // The name must contain at least one character between the starting and ending delimiters.
-        nameLen >= 3 &&
-        (
-           let core = nm.Substring(1, nameLen - 2)
-           // no operator characters except '|' and ' '
-           core |> String.forall (fun c -> c = '|' || c = ' ' || not (opCharSet.Contains c)) &&
-           // at least one non-operator or space character
-           core |> String.exists (fun c -> c = ' ' || not (opCharSet.Contains c))
-        )
+        let nameLen = name.Length
+        if nameLen < 3 || name.[0] <> '|' || name.[nameLen - 1] <> '|' then false else
+
+        let rec isCoreActivePatternName (name: string) idx seenNonOpChar =
+            if idx = name.Length - 1 then seenNonOpChar else
+
+            let c = name.[idx]
+            if opCharSet.Contains(c) && c <> '|' && c <> ' ' then false else
+
+            isCoreActivePatternName name (idx + 1) (seenNonOpChar || c <> '|')
```

Comment:
Stylistically this should probably be:

```fsharp
if idx = name.Length - 1 then
    seenNonOpChar
else
    let c = name.[idx]
    if opCharSet.Contains(c) && c <> '|' && c <> ' ' then
        false
    else
        isCoreActivePatternName name (idx + 1) (seenNonOpChar || c <> '|')
```

Reply:
I find it easier to reason about and debug when there is a single execution path and multiple exit paths. I will reformat the code this way if it's required, though, I'd leave it as is if it's not.

Reply:
@auduchinok,
I personally dislike the indentation style where the else block is inline with the main code path, in my view it hides the return value following the then keyword.

Another thing to remember is that not everyone who looks at these PR's is an expert F# developer, most other developers will be more familiar with the convention where the expression following the if is indented as is the expression following the else.

That said we don't have coding guidelines that encourage or discourage this.  However, if we did, I would certainly push to limit the use of this style.  So please consider re-arranging this block to be a little "in my view" clearer, however, also consider this a suggestion not a merge blocker.

I hope this helps you understand the concerns.

Kevin

Reply:
@KevinRansom Thank you, I do understand the concerns.
It's just that for some particular pieces of code I find the unindented `if` expression style easier to work with.

I've updated the code style as suggested.

Reply:
@KevinRansom Thank you, I do understand the concerns.
It's just that for some particular pieces of code I find the unindented if expression style easier to work with.

I've updated the code style as suggested.